ext/DimensionalDataMakie.jl

```diff
@@ -148,44 +149,47 @@ for (f1, f2) in _paired(:plot => :heatmap, :heatmap, :image, :contour, :contourf
                 Makie.$f2(args...; merged_attributes...)
             end
             # Add a Colorbar for heatmaps and contourf
-            if T isa Real && $(f1 in (:plot, :heatmap, :contourf)) 
+            if T isa Real && $(f1 in (:plot, :heatmap, :contourf))
                 Colorbar(p.figure[1, 2], p.plot;
                     label=DD.label(A), colorbarkw...
                 )
             end
             return p
         end
-        function Makie.$f1!(axis, A::AbstractDimMatrix; 
+        function Makie.$f1!(axis, A::AbstractDimMatrix;
             x=nothing, y=nothing, colorbarkw=(;), attributes...
         )
             replacements = _keywords2dimpairs(x, y)
-            _, _, args, _ = _surface2(A, attributes, replacements)
+            plottrait = Makie.conversion_trait(Makie.Plot{$f2})
+            _, _, args, _ = _surface2(A, plottrait, attributes, replacements)
             # No ColourBar in the ! in-place versions
             return Makie.$f2!(axis, args...; attributes...)
         end
         function Makie.$f1!(axis, A::Observable{<:AbstractDimMatrix};
             x=nothing, y=nothing, colorbarkw=(;), attributes...
         )
             replacements = _keywords2dimpairs(x,y)
-            args =  lift(x->_surface2(x, attributes, replacements)[3], A)
+            plottrait = Makie.conversion_trait(Makie.Plot{$f2})
+            args =  lift(x->_surface2(x, plottrait, attributes, replacements)[3], A)
             p = Makie.$f2!(axis, lift(x->x[1], args),lift(x->x[2], args),lift(x->x[3], args); attributes...)
             return p
         end
     end
 end
 
-function _surface2(A, attributes, replacements)
+function _surface2(A, plottrait, attributes, replacements)
     # Array/Dimension manipulation
     A1 = _prepare_for_makie(A, replacements)
     lookup_attributes, newdims = _split_attributes(A1)
     A2 = _restore_dim_names(set(A1, map(Pair, newdims, newdims)...), A, replacements)
-    args = Makie.convert_arguments(Makie.VertexGrid(), A2)
+
+    args = Makie.convert_arguments(plottrait, A2)
```

Comment:
If you know this is a dimarray and you don't care about expanding dimensions, the thing to do here is to try converting with the plottype first, then get the plottrait and convert using that if the specific conversion didn't work.

Reply:
Ok so we are guaranteed that `A` will be an AbstractDimMatrix. 
For you suggestions do you mean I could just do

```julia
Makie.convert_arguments(Plot{func}, A)
```
and check whether this worked? How will I know if this fails? Will it throw an exception or return a special type?

Reply:
Specifically this code,
https://github.com/MakieOrg/Makie.jl/blob/2b5931ee92c00603a669d02b8103bff49315024b/src/interfaces.jl#L223-L235

which shows how you can check and recurse based on how many times convert_arguments needs to be applied!

Reply:
You would also probably want to `expand_dimensions` first, to get those pesky units out of the way...

Reply:
Instead of recursion you can just call convert_arguments again

Reply:
I tried to match what Makie does more closely, but I am now getting a StackOverflow error when the dimensions have units. 

This no longer works.

```julia
A2u = rand(Y(10u"km":10u"km":100u"km"), X(['a', 'b', 'c']))
heatmap(A2u)
```

I am starting to feel that some of these problems may have to do with the extension overloading the `heatmap`, `plot`, `image`, etc. functions. Is there a more Makie way to do this? I know AxisKeysExtra.jl doesn't seem to need to resort to this kind of machinery https://github.com/JuliaAPlavin/AxisKeysExtra.jl/blob/master/ext/MakieExt.jl
To fix all of these issues, I'm starting to feel the need to reimplement some of Makie's conversion pipeline, which is pretty far beyond my comfort level.

Reply:
Looking into this further, Intervals doesn't play nicely with Unitful and Makie. For example,

```julia
x = 1..3
y = 10u"km"..100u"km"
image(x, y, rand(3, 10))
ERROR: MethodError: no method matching to_endpoints(::Tuple{Quantity{Int64, 𝐋, Unitful.FreeUnits{…}}, Quantity{Int64, 𝐋, Unitful.FreeUnits{…}}})

Closest candidates are:
  to_endpoints(::Any, ::Any)
   @ Makie ~/.julia/packages/Makie/aG6xp/src/conversions.jl:372
  to_endpoints(::IntervalSets.ClosedInterval)
   @ Makie ~/.julia/packages/Makie/aG6xp/src/conversions.jl:368
  to_endpoints(::Union{IntervalSets.Interval, AbstractVector})
   @ Makie ~/.julia/packages/Makie/aG6xp/src/conversions.jl:369
  ...

Stacktrace:
  [1] to_endpoints(x::IntervalSets.ClosedInterval{Quantity{Int64, 𝐋, Unitful.FreeUnits{(km,), 𝐋, nothing}}})
    @ Makie ~/.julia/packages/Makie/aG6xp/src/conversions.jl:368
  [2] to_endpoints(x::IntervalSets.ClosedInterval{Quantity{Int64, 𝐋, Unitful.FreeUnits{(km,), 𝐋, nothing}}}, dim::String)
    @ Makie ~/.julia/packages/Makie/aG6xp/src/conversions.jl:375
  [3] convert_arguments(::ImageLike, xs::IntervalSets.ClosedInterval{…}, ys::IntervalSets.ClosedInterval{…}, data::Matrix{…})
    @ Makie ~/.julia/packages/Makie/aG6xp/src/conversions.jl:422
  [4] convert_arguments(::Type{Image}, ::IntervalSets.ClosedInterval{Int64}, ::Vararg{Any}; kw::@Kwargs{})
    @ Makie ~/.julia/packages/Makie/aG6xp/src/conversions.jl:26
  [5] convert_arguments(::Type{…}, ::IntervalSets.ClosedInterval{…}, ::IntervalSets.ClosedInterval{…}, ::Matrix{…})
    @ Makie ~/.julia/packages/Makie/aG6xp/src/conversions.jl:14
  [6] conversion_pipeline(P::Type{…}, used_attrs::Tuple{}, args::Tuple{…}, args_obs::Tuple{…}, user_attributes::Dict{…}, deregister::Vector{…}, recursion::Int64)
    @ Makie ~/.julia/packages/Makie/aG6xp/src/interfaces.jl:223
  [7] conversion_pipeline(P::Type{…}, used_attrs::Tuple{}, args::Tuple{…}, args_obs::Tuple{…}, user_attributes::Dict{…}, deregister::Vector{…})
    @ Makie ~/.julia/packages/Makie/aG6xp/src/interfaces.jl:214
  [8] (Image)(user_args::Tuple{…}, user_attributes::Dict{…})
    @ Makie ~/.julia/packages/Makie/aG6xp/src/interfaces.jl:272
  [9] _create_plot(::Function, ::Dict{Symbol, Any}, ::IntervalSets.ClosedInterval{Int64}, ::Vararg{Any})
    @ Makie ~/.julia/packages/Makie/aG6xp/src/figureplotting.jl:316
 [10] image(::IntervalSets.ClosedInterval{Int64}, ::Vararg{Any}; kw::@Kwargs{})
    @ MakieCore ~/.julia/packages/MakieCore/fkSJS/src/recipes.jl:436
 [11] image(::IntervalSets.ClosedInterval{Int64}, ::IntervalSets.ClosedInterval{Quantity{…}}, ::Vararg{Any})
    @ MakieCore ~/.julia/packages/MakieCore/fkSJS/src/recipes.jl:434
 [12] top-level scope
    @ REPL[82]:1
Some type information was truncated. Use `show(err)` to see complete types.
```

Reply:
Issue opened https://github.com/MakieOrg/Makie.jl/issues/4176

Reply:
Yeah I just hit this the other day. Thanks for making the issue